runs/ws_logs.py

```diff
@@ -0,0 +1,35 @@
+"""
+This module handles logs received via websockets.
+"""
+import json
+import time
+
+from .models import ContentChannelRun
+from channels.sessions import channel_session
+from django.http import Http404
+
+
+@channel_session
+def connect(message):
+    # Expected path format: /logs/<run_id>/
+    _, run_id = message['path'].strip('/').split('/')
+    message.channel_session['run_id'] = run_id
+    message.reply_channel.send({"accept": True})
+
+
+@channel_session
+def receive(message):
+    run_id = message.channel_session['run_id']
+    try:
+        run = ContentChannelRun.objects.get(run_id=run_id)
+        logfile = run.logfile.path
+    except ContentChannelRun.DoesNotExist:
+        raise Http404
+    record = json.loads(message['text'])
+    timestamp = time.localtime(float(record['created']))
+    timestamp = time.strftime('%Y-%m-%d %H:%M:%S', timestamp)
+    log = '%s - %s - %s - %s - %d - %s' % (
+        record['levelname'], timestamp, record['filename'],
+        record['funcName'], record['lineno'], record['message'])
+    with open(logfile, 'a') as f:
```

Comment:
This look good.

One thing we could possibly do is have three log files for each run:

  - The full, combined log (as above)
  - An errors-only log (if `record['levelname'] == 'ERROR'`)
  - A critical log (if `record['levelname'] == 'CRITICAL'`)

This will require a change to the model, but it would make the frontend simpler (there are alrady three panes for these three types of messages).